coldfront/core/utils/management/commands/audit_data.py

```diff
@@ -0,0 +1,323 @@
+                    f'(it shouldn\'t)'))
+
+    def handle_project_inactive(self):
+        '''
+        Assert that inactive Projects have "Expired" allocations and zero SUs.
+        '''
+        inactive_projects = Project.objects \
+                                .prefetch_related('allocation_set__status',
+                                    'allocation_set__resources',
+                                    'allocation_set__allocationattribute_set') \
+                                .filter(status__name='Inactive') \
+                                .order_by('name')
+
+        for project in inactive_projects:
+            for allocation in project.allocation_set \
+                              .select_related('status') \
+                              .prefetch_related('allocationattribute_set',
+                                                'resources'):
+                try:
+                    resource = allocation.resources.first().name
+                    style = (lambda x:
+                            self.style.WARNING('CURRENTLY EXPECTED BEHAVIOR: ') \
+                            + self.style.WARNING(x)) if \
+                            resource.endswith('Directory') \
+                            else self.style.ERROR
+                    if allocation.status.name != 'Expired':
+                        self.stdout.write(style(f'Project {project.name} is '
+                        f'inactive and has an unexpired '
+                        f'{resource} allocation {allocation.id}.'))
+                except AttributeError as e:
+                    pass
+                try:
+                    allocation_attribute = allocation.allocationattribute_set \
+                        .get(allocation_attribute_type__name='Service Units')
+                    if float(allocation_attribute.value) != 0:
+                        self.stdout.write(self.style.ERROR(
+                        f'Project {project.name} is inactive and has '
+                        f'non-zero SUs.'))
+                except AllocationAttribute.DoesNotExist as e:
+                    pass
+
+    def handle_project_pi(self):
+        '''
+        Assert that all projects have at least one PI.
+        '''
+        projects = Project.objects \
+            .select_related('status') \
+            .order_by('status__name', 'name')
+        for project in projects:
+            if not project.pis().exists():
+                self.stdout.write(self.style.ERROR( f'{project.status.name} '
+                f'Project {project.name} has no PIs.'))
+
+    def handle_user_project(self):
+        '''
+        Assert that all users with a cluster UID should be associated with
+        at least one Project.
+        '''
+        users = UserProfile.objects.filter(cluster_uid__isnull=False) \
+            .select_related('user') \
+            .order_by('user__is_active', 'user__username') \
+            .values('id', 'user__first_name', 'user__last_name',
```

Comment:
Is there a reason to fetch and use the values in this way?

Instead of `user["user__last_name"]`, you can do `user.user.last_name`. (The first `user` would be more accurately named `user_profile`.)

If the concern is database lookups, you can fetch everything you need all upfront using `prefetch_related` and `select_related`.

Reply:
I heard that it's more efficient but if that doesn't matter I can swap it to the normal usage

Reply:
Please do, mainly for consistency with the rest of the codebase.

I've never used [`only`](https://docs.djangoproject.com/en/4.1/ref/models/querysets/#only), but it looks like it accomplishes the same thing (it only retrieves the fields you want), while still returning ORM objects.

```
>>> User.objects.first().__dict__
{'_state': <django.db.models.base.ModelState object at 0x7f9580a07128>, 'id': 1, 'password': '', 'last_login': None, 'is_superuser': False, 'username': 'username', 'first_name': 'First', 'last_name': 'Last', 'email': 'email@berkeley.edu', 'is_staff': False, 'is_active': True, 'date_joined': datetime.datetime(1970, 1, 1, 0, 0, 0, 0, tzinfo=<UTC>)}
>>> User.objects.only("first_name").first().__dict__
{'_state': <django.db.models.base.ModelState object at 0x7f9580a07898>, 'id': 1, 'first_name': 'First'}
```